File: 2614-maximum-count-of-positive-integer-and-negative-integer/2614-maximum-count-of-positive-integer-and-negative-integer.java
class Solution {
    public int lastNegative(int[] arr, int n) {
        int low = 0, high = n - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] < 0) {
                if (mid == n - 1 || arr[mid + 1] >= 0) 
                    return mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return -1;
    }

    public int firstPositive(int[] arr, int n) {
        int low = 0, high = n - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] > 0) {
                if (mid == 0 || arr[mid - 1] <= 0)  // Fixed condition
                    return mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }

    public int maximumCount(int[] arr) {
        int n = arr.length;
        int x = lastNegative(arr, n);
        int y = firstPositive(arr, n);

        int neg = x + 1;  // Count of negative numbers
        int pos = (y == -1) ? 0 : n - y;  // Count of positive numbers

        return Math.max(pos, neg);
    }

};

File: 2614-maximum-count-of-positive-integer-and-negative-integer/2614-maximum-count-of-positive-integer-and-negative-integer.cpp
class Solution {
public:
    int BS1(vector<int>& nums, int n) {
        int low = 0, high = n - 1, mid;

        while (low <= high) {
            mid = (low + high) / 2;

            if (mid == n - 1 || (nums[mid] < 0 && nums[mid + 1] >= 0))
                return mid;
            else if (nums[mid] < 0)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return -1;
    }

    int BS2(vector<int>& nums, int n) {

        int low = 0, high = n - 1, mid;

        while (low <= high) {
            mid = (low + high) / 2;

            if (mid == 0 || (nums[mid] > 0 && nums[mid - 1] <= 0))
                return mid;
            else if (nums[mid] > 0)
                high = mid - 1;
            else
                low = mid + 1;
        }
        return -1;
    }

    int maximumCount(vector<int>& nums) {
        int n = nums.size();

        if(n == 1 && nums[0] == 0)
            return 0;

        int x = BS1(nums, n);
        int y = BS2(nums, n);

        int neg = x + 1;
        int pos = (y == -1)? 0 : n - y;

        return max(pos, neg);
    }
};